### ISSUE TITLE
Intro to Testing and Test Driven Development

### ISSUE THREAD

**Issue by acviana** [STScI-Citizen-Science/MTPipeline#79]
### Summery

A core principle in software development is testing. Testing a broad area with many different methodologies, proponents, detractors, etc. This issue is designed to introduce you to what is called unit testing in Python using the `nosetests` module and a test-driven development methodology.

**Comment by acviana** [comment]
To start with, let's create a new environment for this issue. Based on our last ticket, create a new virtual environment using `virtualenv` under `ureka` and call it `completeness-test`. In this environment ensure that the latest version of the `nose` package is installed and clone a copy of the MTPipeline repo. Show me the commands you used to do this.

**Comment by walyssonBarbosa** [comment]
I used the commands below to do the new environment:

```
$ ur_setup
$ cd virtualenvs/
$ virtualenv completeness-test
$ cd completeness-test
$ source bin/activate.csh
```

To install the latest version of the `nose` package:

```
$ pip install nose --upgrade
```

To clone and copy the repo:

```
$ git clone https://github.com/STScI-Citizen-Science/MTPipeline.git
```

And then:

```
$ cd MTPipeline/
$ pip install numpy
$ python setup.py develop
```

**Comment by acviana** [comment]
:tada: Good Job! :tada:  Let's put that environment aside for now, we'll come back to it. 

### FizzBuzz

FizzBuzz is an entry-level programming question common used as a basic interview question. The statement problem is to construct a function that has some variation on the following behavior:
- The function takes an integer
- If the integer isn't divisible by either 3 or 5 it returns the integer
- If the integer is only divisible by 3 return the string 'fizz'
- If the integer is only divisible by 5 return the string 'buzz'
- If the integer is divisible by 3 and 5 return the string 'fizzbuzz'

So normally you would just write the function and be done with it. But we're going to use this as an exercise to learn test-driven development (TDD) for unit testing. In this case the `fizzbuzz` function you are going to write is the "unit" we are testing. And in test drive development, you write the test _first_, and then write a function that passes the test.

So create a folder (outside the repo) called `fizzbuzz` and create two python files inside it called `fizzbuzz.py` and `test_fizzbuzz.py`. Then read the `nosetests` docs and create a function called `test_fizzbuzz` in `test_fizzbuzz.py` that you think covers all the distinct output cases for a theoretical `fizzbuzz` function. Note that we haven't written the `fuzzbuzz` function yet, and we don't need to because we can already describe it's desired behavior. Post the code below once you're done.

**Comment by walyssonBarbosa** [comment]
This would be the function:

``` python
def test_fizzbuzz():
    assert fizzbuzz(1) == 1
    assert fizzbuzz(3) == "fizz"
    assert fizzbuzz(10) == "buzz"
    assert fizzbuzz(45) == "fizzbuzz"
```

The function `fizzbuzz` used here should be implemented in `fizzbuzz.py`.

**Comment by acviana** [comment]
Great! Here's what's next:
- Add an error message to your assert statements to help debug any errors. Something like this: 

``` python
assert fizzbuzz(1) == 1, expected 1 got {}'.format(fizzbuzz(1))
```
- Add this line at the top of `test_fizzbuzz.py`: `from fizzbuzz import fizzbuzz`
- Add a dummy function to `fizzbuzz.py` like this:

``` python
def fizzbuzz(integer):
    pass
```
- Now run the tests with this command in the same folder as the test module. I should say you ran 1 test and it failed. 

``` shell
$ nosetests ﻿test_fizzbuzz.py
```
- Now update the `fizzbuzz.py` function until it passes the test.

**Comment by walyssonBarbosa** [comment]
Should I post the code here?

**Comment by acviana** [comment]
Yes, please.

**Comment by walyssonBarbosa** [comment]
Here is the code for `fizzbuzz.py`:

``` python
def fizzbuzz(integer):
    if(integer % 15 == 0):
        return "fizzbuzz"
    elif(integer % 5 == 0):
        return "buzz"
    elif(integer % 3 == 0):
        return "fizz"
    else:
        return integer
```

And the code for `test_fizzbuzz.py`:

``` python
from fizzbuzz import fizzbuzz

def test_fizzbuzz():
    assert fizzbuzz(1) == 1, "expected 1 got {}".format(fizzbuzz(1))
    assert fizzbuzz(3) == "fizz", "expected fizz got {}".format(fizzbuzz(3))
    assert fizzbuzz(5) == "buzz", "expected buzz got {}".format(fizzbuzz(5))
    assert fizzbuzz(15) == "fizzbuzz", "expected fizzbuzz got {}".format(fizzbuzz(15))
```

**Comment by acviana** [comment]
Great! One comment, in Python you don't need to put your control statements in parenthesis the way you did. Python supports this because it's liberal with extra parenthesis, but I suspect this is a Java syntax pattern.

### Testing with Generators

So we already have a test and a function that passes that test so we are done with the original assignment. But before we close this ticket I would like you to work on a variation of your original code. Right now you entire `test_fizzbuzz` function is counted as a single test, even though it tests 4 different conditions. This is kind of up to personal preference but there are times when you might want to count each assert as a different test. 

There are two ways to do this I can think of. The trivial case is just to make 4 different testing functions each of which only tests a single condition. We don't need to go through that exercise. The other way is to use test generators. In essence a generator uses a `yield` statement to iteratively return values on demand as opposed to a `return` statement which returns everything all at once. This is a useful feature is certain computationally or memory-bound problems. You can read about generators in this blog post by Jeff Knupp: http://www.jeffknupp.com/blog/2013/04/07/improve-your-python-yield-and-generators-explained/

The basic design pattern I use for writing test generators is outlined in the `nosetests` documentation: https://nose.readthedocs.org/en/latest/writing_tests.html#test-generators. Copying directly from the docs: 

``` python
def test_evens():
    for i in range(0, 5):
        yield check_even, i, i*3

def check_even(n, nn):
    assert n % 2 == 0 or nn % 2 == 0
```

Basically, you define some generalized "checking" function that performs the actual `assert` command and then your "test" function iterates over some loop, yielding an instance of the check function on each iteration. `nosetests` will consider each instance of the checking function yielded by testing function a separate test. So update you original code to this pattern to run the same 4 assert statements but this time use this generator pattern to produce 4 tests instead of 1.

**Comment by walyssonBarbosa** [comment]
This would be the code for `test_fizzbuzz.py` using a test generator:

``` python
from fizzbuzz import fizzbuzz

def test_fizzbuzz():
    for i in [[1, 1], [3, "fizz"], [5, "buzz"], [15, "fizzbuzz"]]:
        yield check_fizzbuzz, i[0], i[1]

def check_fizzbuzz(integer, expected_output):
    assert fizzbuzz(integer) == expected_output, "expected {0} got {1}".format(expected_output, fizzbuzz(integer))
```

**Comment by acviana** [comment]
Good! A few suggestions, let's look at `test_fizzbuzz`. I agree with your decision to use a list as the iterable for your loop. However, I wouldn't make the elements of that list lists themselves. My reasoning for this is that lists are used when _order_ matters and dictionaries are used when _names_ matter. In this case you are using the 0th and 1st position to differentiate the "integer" and "expected_output" test values but you don't really care about the position, you care about the names. Instead, I think it would be clearer to make a list of _dictionaries_ with descriptive keys such as the ones I mentioned. I think that little design decisions like this go a long way in making code more readable and maintainable.

**Comment by walyssonBarbosa** [comment]
Would it be something like this?

``` python
from fizzbuzz import fizzbuzz

input_values = {1: 1, 6: "fizz", 10: "buzz", 45: "fizzbuzz"}

def test_fizzbuzz():
    for i in input_values:
        yield check_fizzbuzz, i, input_values[i]

def check_fizzbuzz(integer, expected_output):
    assert fizzbuzz(integer) == expected_output, "expected {0} got {1}".format(expected_output, fizzbuzz(integer))
```

**Comment by acviana** [comment]
Does that run? I was thinking something more like this:

``` python
from fizzbuzz import fizzbuzz

def test_fizzbuzz():
    input_list = [{'integer':1, 'expected_output':1}, 
                  {'integer':6, 'expected_output':'fizz'}, 
                  {'integer':10, 'expected_output':'buzz'}, 
                  {'integer':45, 'expected_output':'fizzbuzz'}]
    for input_dict in input_lise:
        yield check_fizzbuzz, input_dict['integer'], input_dict['expected_output']

def check_fizzbuzz(integer, expected_output):
    assert fizzbuzz(integer) == expected_output, "expected {0} got {1}".format(expected_output, fizzbuzz(integer))
```

**Comment by walyssonBarbosa** [comment]
Yes, it does. But this one is clearer than my implementation.

**Comment by acviana** [comment]
Ok, unless you have any lingering questions I think we can close this one.

:clap: :clap: :clap:

**Comment by walyssonBarbosa** [comment]
Okay!

I'm back, by the way.
